crates/ruff_python_semantic/src/node.rs

```diff
@@ -0,0 +1,105 @@
+use std::num::TryFromIntError;
+use std::ops::{Index, IndexMut};
+
+use rustc_hash::FxHashMap;
+use rustpython_parser::ast::Stmt;
+
+use ruff_python_ast::types::RefEquality;
+
+/// Id uniquely identifying a statement in a program.
+///
+/// Using a `u32` is sufficient because Ruff only supports parsing documents with a size of max `u32::max`
+/// and it is impossible to have more statements than characters in the file.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct NodeId(u32);
```

Comment:
I wonder if we should use `NonZeroU32` here (and either offset the `indices` when accessing the `nodes` vec or add a dummy node to the vec). The benefit of doing so is that it would shrink the `Option<NodeId>` from 8 bytes to 4 bytes. 

Or we introduce a new, module internal, `ParentId` that uses `NonZeroU32` internally that can be converted from/to a `NodeId`. This requires some more boilerplate but may be easier to implement (but limits the benefits of this optimization to this module only).

Reply:
Gave it a shot!